runtime/job-workers/shared-memory-manager.cpp

```diff
@@ -5,54 +5,132 @@
-
-  for (size_t i = 0; i < slices_count_; ++i) {
-    auto *slice = get_message(i);
-    slice = new(slice) JobSharedMessage();
-    slice->resource.init(reinterpret_cast<uint8_t *>(slice) + control_block_size, memory_slice_size_ - control_block_size);
+  assert(!messages_);
+  assert(!owners_table_);
+
+  const size_t processes_count = vk::singleton<job_workers::JobWorkersContext>::get().job_workers_num + std::max(workers_n, 1);
+  if (!memory_limit_) {
+    memory_limit_ = processes_count * 8 * 1024 * 1024;
   }
+  auto *raw_mem = static_cast<uint8_t *>(mmap(nullptr, sizeof(SharedMemoryProcessOwnersTable) + memory_limit_,
+                                              PROT_READ | PROT_WRITE, MAP_SHARED | MAP_ANONYMOUS, -1, 0));
+  assert(raw_mem);
+  owners_table_ = new(raw_mem) SharedMemoryProcessOwnersTable{};
+  raw_mem += sizeof(SharedMemoryProcessOwnersTable);
+
+  size_t left_memory = memory_limit_;
+  const size_t messages_count_limit = 2 * (vk::singleton<job_workers::JobWorkersContext>::get().job_workers_num + std::max(workers_n, 1));
+  messages_count_ = std::min(messages_count_limit, left_memory / sizeof(JobSharedMessage));
+  messages_ = new(raw_mem) JobSharedMessage[messages_count_];
 }
 
 JobSharedMessage *SharedMemoryManager::acquire_shared_message() noexcept {
-  const size_t random_slice_idx = rd_() % slices_count_;
+  assert(messages_);
+  const size_t random_slice_idx = rd_() % messages_count_;
   size_t i = random_slice_idx;
 
   do {
-    auto *slice = get_message(i);
-    pid_t prev_pid = 0;
-    if (slice->owner_pid.compare_exchange_strong(prev_pid, pid)) {
-      JobStats::get().currently_memory_slices_used++;
-      return slice;
+    auto *message = messages_ + i;
+    uint32_t expected_ref_cnt = JOB_SHARED_MESSAGE_FREE_CNT;
+    {
+      dl::CriticalSectionGuard critical_section;
+      if (message->owners_counter.compare_exchange_strong(expected_ref_cnt, 1)) {
+        owners_table_->workers_table[logname_id].attach(message);
+        ++JobStats::get().currently_messages_acquired;
+        return message;
+      } else {
+        ++JobStats::get().message_acquire_false_iterations;
+      }
     }
 
-    if (++i == slices_count_) {
+    if (++i == messages_count_) {
       i = 0;
     }
   } while (i != random_slice_idx);
   return nullptr;
 }
 
-void SharedMemoryManager::release_shared_message(JobSharedMessage *slice) noexcept {
-  hard_reset_var(slice->instance);
-  slice->resource.init(slice->resource.memory_begin(), slice->resource.get_memory_stats().memory_limit);
-  slice->job_id = 0;
-  slice->job_result_fd_idx = -1;
-  const pid_t prev_pid = slice->owner_pid.exchange(0);
-  php_assert(prev_pid);
-  JobStats::get().currently_memory_slices_used--;
+void SharedMemoryManager::release_shared_message(JobSharedMessage *message) noexcept {
+  dl::CriticalSectionGuard critical_section;
+  owners_table_->workers_table[logname_id].detach(message);
+  if (--message->owners_counter == 0) {
```

Comment:
`message->owners_counter` is always 1 or 0. I think it would be more logical to consider it as bool flag `message->released` and set it to `false` here instead of decrement